### PR TITLE
BUG: pivot_table not returning correct type when margin=True and aggfunc='mean' 

### LINKED ISSUE #24893: pivot_table margins=True default aggfunc='mean' does integer division
#### Code Sample, a copy-pastable example if possible

Code
```
df = pd.DataFrame({'State':['Texas', 'Texas', 'Florida', 'Florida'], 
                   'a':[4,5,1,3], 'b':[6,10,3,11]},index=['one','two','three','four'])
pd.pivot_table(df,index='State',margins=True)
```
Output
```
	a	b
State		
Florida	2.00	7
Texas	4.50	8
All	3.25	7
```
#### Problem description

The margin value for the `b` column is 7 (calculated from default aggfunc mean of 7+8) when i expect 7.5. 

This issue seems to be opposite of issue #17013 which complains of integers becoming floats 

#### Expected Output
P.s I changed 6 to 6.0 in to create desired output.
Code
```
df = pd.DataFrame({'State':['Texas', 'Texas', 'Florida', 'Florida'], 
                   'a':[4,5,1,3], 'b':[6.0,10,3,11]},index=['one','two','three','four'])
pd.pivot_table(df,index='State',margins=True)
```
Desired Output (For the margin value only, i don't necessarily require 7, 8 to become 7.0, 8.0)
```
	a	b
State		
Florida	2.00	7.0
Texas	4.50	8.0
All	3.25	7.5

```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.0.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.4
pytest: 3.8.0
pip: 18.1
setuptools: 40.2.0
Cython: 0.28.5
numpy: 1.15.4
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: 1.7.9
patsy: 0.5.0
dateutil: 2.7.3
pytz: 2018.5
blosc: None
bottleneck: 1.2.1
tables: 3.4.4
numexpr: 2.6.8
feather: None
matplotlib: 3.0.2
openpyxl: 2.5.6
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.1.0
lxml: 4.2.5
bs4: 4.6.3
html5lib: 1.0.1
sqlalchemy: 1.2.11
pymysql: None
psycopg2: 2.7.6.1 (dt dec pq3 ext lo64)
jinja2: 2.9.5
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.7.0
</details>

### PR DESCRIPTION
**Pull request by mabelvj**
- [x] closes #24893
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

# What's new
Use of `maybe_downcast_to_dtype` in `_add_margins` so it can resolve the dtype conversion, avoiding floats being converted to integers when the result of the `aggfunc` is a float.

For the new case, if after applying the aggfunc,  if the margin result is not an integer, the whole column is converted to float:

Example:

```
_df
   A  B  C  D
0  2  1  1  X
1  4  4  3  X
2  6  5  4  Y
3  8  8  6  Y
```

Currently pandas does this:

```
print(pd.pivot_table(_df, index="D", margins=True))                                                              
     A    B  C
D             
X    3  2.5  2
Y    7  6.5  5
All  5  4.5  3

```

with the fix the result is:

```
print(pd.pivot_table(_df, index="D", margins=True))                                                              
     A    B  C
D             
X    3  2.5  2.0
Y    7  6.5  5.0
All  5  4.5  3.5

```

# Issues
There are test referencing that np.means of ints are casted back into ints. However, giving that for the aggregations in the rows, floats are kept when the np.mean of integers is a float, it does not make sense that this behavior does not hold for the margins.

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
@mabelvj Thanks for the PR.

to fix the ci failure you'll need to run `black pandas` https://dev.pandas.io/development/contributing.html#python-pep8-black 

also test to confirm the fix works.

**Comment by mabelvj** [comment]
I see that there are tests referencing that np.means of ints are casted back into ints marked with `pytest.mark.xfail` that are related to this issue, should I update them?:
https://github.com/pandas-dev/pandas/blob/03b3c8fc82b3a18a3ddcad1b3b26d601467fc74c/pandas/tests/reshape/test_pivot.py#L1604
https://github.com/pandas-dev/pandas/blob/03b3c8fc82b3a18a3ddcad1b3b26d601467fc74c/pandas/tests/reshape/test_pivot.py#L1618

**Comment by TomAugspurger** [comment]
> should I update them?

Has the output changed? I would have expected them to fail CI if they were changed to no longer cast from float to int.

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. can you add a release note in reshaping bug fix section for 1.0

File: pandas/core/reshape/pivot.py [hunk 1]
```diff
@@ -254,7 +254,9 @@ def _add_margins(
     try:
```
Comment (jreback) [inline pandas/core/reshape/pivot.py]: can you add a comment here about what is going on

**Comment by mabelvj** [comment]
@TomAugspurger the results do not change for those, the tests are skipped if failing since they are marked to do so. Still, checking them (since they are related to this issue, because they check the float casting on the margins),  the expected values do not match the output result (old and new pivot_table give the same output here).

File: pandas/tests/reshape/test_pivot.py [hunk 2]
```diff
@@ -1630,6 +1630,25 @@ def test_categorical_margins_category(self, observed):
...
+                "D": ["X", "X", "Y", "Y"],
+            }
+        )
+
+        result = pd.pivot_table(df, index="D", margins=True)
+        expected = pd.DataFrame(
+            {"A": [3, 7, 5], "B": [2.5, 6.5, 4.5], "C": [2, 5, 3.5]},
+            index=pd.Index(["X", "Y", "All"], name="D"),
+        )
+        table = result
```
Comment (WillAyd) [inline pandas/tests/reshape/test_pivot.py]: This line is causing the CI failure - should be able to just remove

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Review by TomAugspurger — APPROVED** [review]

**Comment by jreback** [comment]
this lgtm. @mabelvj can you merge master; ping on green.

**Comment by WillAyd** [comment]
ping @mabelvj - if you can fix up merge conflicts can get this one in

**Comment by mabelvj** [comment]
@jreback @WillAyd Fixed the conflicts, but Flake8 keeps giving a false positive F841 error on [L1670-1671](https://github.com/mabelvj/pandas/blob/dd121dac85bdbdf5d0552e3692f3f75c498e761c/pandas/tests/reshape/test_pivot.py#L1670-L1671) because of unused variables, but they are used at the last line of the function.

It's strange because other functions have the same structure and do not raise any error.

**Comment by WillAyd** [comment]
I think you need to add `tm.assert_result_equal(result, expected)` to the end of the test

**Comment by mabelvj** [comment]
@WillAyd Fixed now. It seems the line disappeared when fixing previous conflicts with master.

**Review by WillAyd — APPROVED** [review]

**Comment by WillAyd** [comment]
Great - thanks @mabelvj for seeing this one through!

**Pull request merged**